### PR TITLE
Add `deletionDate` to LocalAccount

### PR DESCRIPTION
**Pull request by Milena-Czierlinski**
# Readiness checklist

-   [x] I added/updated tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.
-   [x] I self-reviewed the PR.

### CODE REVIEW

File: packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts [hunk 1]
```diff
@@ -33,7 +34,7 @@ export class DatawalletSynchronizedModule extends AppRuntimeModule<DatawalletSyn
 
         switch (mostRecentIdentityDeletionProcess.status) {
             case IdentityDeletionProcessStatus.Approved:
-                newDeletionDate = mostRecentIdentityDeletionProcess.gracePeriodEndsAt;
+                newDeletionDate = CoreDate.from(mostRecentIdentityDeletionProcess.gracePeriodEndsAt!);
```
Comment (jkoenig134) [inline packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts:37]: hmm, why do we need the null-check here? And is the null-check safe, b/c it's in status Approved?

**Review by jkoenig134 — CHANGES_REQUESTED** [review]
some details, looks very good overall!

File: packages/app-runtime/src/multiAccount/data/LocalAccount.ts [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
-    id: CoreId;
-    address?: CoreAddress;
+    id: ICoreId;
+    address?: ICoreAddress;
     name: string;
     directory: string;
     order: number;
     lastAccessedAt?: ICoreDate;
     devicePushIdentifier?: string;
+    deletionDate?: ICoreDate;
```
Comment (Magnus-Kuhn) [inline packages/app-runtime/src/multiAccount/data/LocalAccount.ts:12]: Why only the deletion date and not the status too?

Comment (jkoenig134) [reply packages/app-runtime/src/multiAccount/data/LocalAccount.ts:12]: b/c we don't care. 

1. deletionDate set = in deletion
2. deletionDate not set = not in deletion

File: packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts (hunk 1 above)
Comment (Milena-Czierlinski) [reply packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts:37]: `gracePeriodEndsAt` is an optional property, that is only set if the status is Approved. We know that, but the compiler doesn't. I suppose we could add type guards for the IdentityDeletionProcess, but I'm not sure if it's worth the effort.

Comment (jkoenig134) [reply packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts:37]: Ok, that's all I wanted to know! Great!

File: packages/app-runtime/src/modules/runtimeEvents/IdentityDeletionProcessStatusChangedModule.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                break;
+
+            case IdentityDeletionProcessStatus.Cancelled:
+                const account = await this.runtime.multiAccountController.getAccountByAddress(event.eventTargetAddress);
+                const previousDeletionDate = account.deletionDate;
+
+                if (!previousDeletionDate) break;
+
+                await this.runtime.multiAccountController.updateLocalAccountDeletionDate(event.eventTargetAddress, undefined);
+                break;
```
Comment (Magnus-Kuhn) [inline packages/app-runtime/src/modules/runtimeEvents/IdentityDeletionProcessStatusChangedModule.ts:34]: no publishing of a `LocalAccountDeletionDateChangedEvent`?

File: packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        const services = await this.runtime.getServices(event.eventTargetAddress);
+        const identityDeletionProcessResult = await services.transportServices.identityDeletionProcesses.getIdentityDeletionProcesses();
+
+        if (identityDeletionProcessResult.isError) {
+            this.logger.error(identityDeletionProcessResult);
+            return;
+        }
+
+        let newDeletionDate;
+        const mostRecentIdentityDeletionProcess = identityDeletionProcessResult.value[identityDeletionProcessResult.value.length - 1];
```
Comment (Magnus-Kuhn) [inline packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts:33]:
```suggestion
        const mostRecentIdentityDeletionProcess = identityDeletionProcessResult.value.at(-1)
```

File: packages/app-runtime/src/multiAccount/MultiAccountController.ts [hunk 5]
```diff
@@ -238,6 +251,22 @@ export class MultiAccountController {
...
+        if (!oldAccount) {
+            throw TransportCoreErrors.general.recordNotFound(LocalAccount, address).logWith(this._log);
+        }
+
+        const account = LocalAccount.from(oldAccount);
+
+        account.deletionDate = deletionDate ?? undefined;
+        await this._localAccounts.update(oldAccount, account);
+
+        const cachedAccount = this.sessionStorage.findSession(address)?.account;
```
Comment (Milena-Czierlinski) [inline packages/app-runtime/src/multiAccount/MultiAccountController.ts:266]: @jkoenig134 Should this always be defined? (I would like to return the updated cachedAccount, ideally knowing that it isn't undefined.)

Comment (jkoenig134) [reply packages/app-runtime/src/multiAccount/MultiAccountController.ts:266]: I think you could run "findOrCreateSession" IDK if it's called like this to make sure that it is defined. But I'd be careful about this cache and not rely on it too much.

File: packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts (hunk 4 above)
Comment (jkoenig134) [reply packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts:33]: I think we should use this at many more places (in another refactoring PR :) )

File: packages/app-runtime/src/modules/runtimeEvents/IdentityDeletionProcessStatusChangedModule.ts (hunk 3 above)
Comment (jkoenig134) [reply packages/app-runtime/src/modules/runtimeEvents/IdentityDeletionProcessStatusChangedModule.ts:34]: nah, not needed when the change was made by the current device

File: packages/transport/src/modules/accounts/AccountController.ts [hunk 6]
```diff
@@ -230,11 +230,9 @@ export class AccountController {
...
 
     public async syncDatawallet(force = false): Promise<void> {
-        if (!force && !this.autoSync) {
-            return;
-        }
+        if (!force && !this.autoSync) return;
 
-        return await this.synchronization.sync("OnlyDatawallet");
+        await this.synchronization.sync("OnlyDatawallet");
     }
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/src/modules/accounts/AccountController.ts:236]: that's a VERY bad advice and goes against everything we work for making the codebase cleaner.

File: packages/app-runtime/test/modules/DatawalletSynchronizedModule.test.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+        await sessionDevice2.transportServices.account.syncDatawallet();
+        const event = await TestUtil.awaitEvent(runtimeDevice2, LocalAccountDeletionDateChangedEvent);
+        expect(event.data).toBe(initiateDeletionResult.value.gracePeriodEndsAt);
+
+        expect(sessionDevice2.account.deletionDate).toBe(initiateDeletionResult.value.gracePeriodEndsAt);
+
+        const account = await runtimeDevice2.multiAccountController.getAccount(CoreId.from(sessionDevice2.account.id));
+        expect(account.deletionDate!.toString()).toBe(initiateDeletionResult.value.gracePeriodEndsAt);
+    });
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/app-runtime/test/modules/DatawalletSynchronizedModule.test.ts:62]: It's a double check for persistence

File: packages/app-runtime/src/multiAccount/MultiAccountController.ts [hunk 8]
```diff
@@ -238,6 +251,22 @@ export class MultiAccountController {
...
+        if (cachedAccount) cachedAccount.deletionDate = deletionDate?.toString();
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/app-runtime/src/multiAccount/MultiAccountController.ts:267]: yeah, we actually do in our tests.

[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/app-runtime/src/multiAccount/MultiAccountController.ts:267]: That's exactly what your `What's the difference between this account and the account above?` question is about.

File: packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts (hunk 4 above)
Comment (jkoenig134) [reply packages/app-runtime/src/modules/runtimeEvents/DatawalletSynchronizedModule.ts:33]: ^ done in #348

**Review by jkoenig134 — APPROVED** [review]
LGTM

**Comment by Milena-Czierlinski** [comment]
@Siolto Please note that the DatawalletSynchronizedEvent is now published in transport instead of the app-runtime, which is why its namespace changes from `"app.datawalletSynchronized"` to `"transport.datawalletSynchronized"`.

**Comment by jkoenig134** [comment]
@Magnus-Kuhn please re-check your comments and resolve if you are ok with the answers. If not we have to discuss the things.

**Pull request merged**